Add persistent_term:put_new

Benjamin-Philip opened the pull request:
This PR adds the persistent_term:put_new to conditionally add a key-value only if the key is new.

Closes #9681.

/cc @jhogberg.

josevalim commented:
Instead of adding `put_new`, would it be more useful to add a `has_key` function that could be used in more cases? I recognize `put_new` has the benefit of being atomic but I assume that the value you need to store is likely expensive to compute anyway, so you want to check if the key exists before you do any work.

Benjamin-Philip commented:
I think a `has_key` would be a good addition, but not as a replacement to `put_new`. Like you said `put_new` has the benefit of being atomic, enabling one to handle data with duplicate keys parallelly in a single pass. See the linked issue for my use case.

You can define a rudimentary `has_key` already by setting `Default` in `get/2` to `undefined` or some other more specific atom. Nevertheless, `has_key` would be faster and less error prone.

Benjamin-Philip commented:
@jhogberg, I think this PR is review ready now. Sorry about the delay - I was busy the past 2 weeks...

Benjamin-Philip commented:
Rebased on master to fix merge conflicts.

jhogberg commented:
> @jhogberg, Any updates on this? Do you think you can squeeze it in OTP 28?

As mentioned in the issue, the window for 28 was closed before you opened it. It can be included in 28.1 at the very earliest.

The refactor I mentioned is coming along well. It's feature-complete and offers far better add/update/erase performance (so long as literal GC isn't triggered), but read performance on dynamic keys has taken a hit that is proving a bit annoying to get rid of (statically known keys are effectively free however). I'm confident it'll work out however, and am aiming for OTP 29.

>  Also, do you think `has_key` should be added (in another PR, of course) as @josevalim suggested?

I can't see why not, I'll add it to my refactored version.

Benjamin-Philip commented:
> It can be included in 28.1 at the very earliest.

When do you think 28.1 will be released?

jhogberg commented on erts/emulator/beam/erl_bif_persistent.c:
```diff
@@ -395,49 +389,101 @@ BIF_RETTYPE persistent_term_put_2(BIF_ALIST_2)
+    ERTS_BIF_YIELD_RETURN(BIF_P, am_ok);
+}
+
+BIF_RETTYPE persistent_term_put_new_2(BIF_ALIST_2)
```
I think this shares a bit too much with `put_2`, why not merge them both into a separate function with the relevant `Export*` (for trapping) and a flag to control the behavior on key clashes?

Benjamin-Philip commented:
> 28.1 would be maint, it should be relatively easy to rebase to that though since master hasn't moved in this area. :-)

I've rebased to maint and changed the PR base from master to maint as well.

> I've paused work on it because more important things came up. The idea was to use [concurrent tries](https://en.wikipedia.org/wiki/Ctrie) to improve update performance for trivial values, sacrificing a bit of read performance but hopefully regaining it through inline caching. Inline caching of static keys resulted in great performance (marginally more expensive than reading a list head...), but I couldn't make dynamic keys fast enough.
>
> I left it just before introducing namespacing, replacing the implicit persistent_term:put({?MODULE, Key}, Value) with an explicit persistent_term:put(?MODULE, Key, Value) where the user could define how they wanted their namespace to work -- slow writes but fast reads through the old hashtable method, or slightly slower reads but fast writes through concurrent tries.

I think namespacing is a great idea - you reduce module, key pairs into a trivial atom. If you're creating a new hashtable/ctrie for each namespace you also get some additional isolation between namespaces in terms of failures and maybe even GC?

Regarding static and dynamic keys, do you mean keys known at compile time? Maybe we could cache all trivial keys (i.e. a key that is an atom)? This may not scale well with persistent terms as they are now, but if caching like this for each namespace was configured intentionally on a case by case it may work quite well.

Benjamin-Philip commented:
> It looks like you dragged some unrelated commits into it.

I rebased and forced pushed *before* updating the PR's base from master to maint. So GitHub dragged commits from maint that weren't on master.

I'm going to make a new branch and cherry-pick my commits on to it and update the PR branch.

Benjamin-Philip commented on erts/emulator/test/persistent_term_SUITE.erl:
```diff
@@ -960,12 +984,18 @@ put_erase_trapping(_Config) ->
     Value = {ValuePrefix, NrOfPuts},
-    erts_debug:set_internal_state(reds_left, rand:uniform(250)),
-    persistent_term:put(Key, Value),
+    %% erts_debug:set_internal_state(reds_left, rand:uniform(250)),
```
For some reason the trapping tests fail due to a badarg from `get/1` when this is uncommented. This badarg seems to magically disappear when this line is commented.

I don't understand why changing the number of reductions left would cause `put/2` and `put_new/2` to effectively not execute. Does this point towards some error in my trapping?

Also I don't understand why this error only showed up in my last change and not in any earlier ones.

Benjamin-Philip replied:
It's these lines:

```erlang
erts_debug:set_internal_state(reds_left, rand:uniform(250)),
erlang:apply(persistent_term, PutFun, [Key, Value]),
Value = persistent_term:get(Key),
```

`get/1` fails with the `badarg`, so I'm assuming that's because `put/2` and `put_new/2` don't actually update.

jhogberg commented on erts/emulator/test/persistent_term_SUITE.erl:
```diff
@@ -960,12 +984,18 @@ put_erase_trapping(_Config) ->
+    %% erts_debug:set_internal_state(reds_left, rand:uniform(250)),
+    erlang:apply(persistent_term, PutFun, [Key, Value]),
     Value = persistent_term:get(Key),
     do_puts(NrOfPuts - 1, ValuePrefix).
```
```suggestion
    `do_puts_iter(NrOfPuts - 1, ValuePrefix, PutFun).`
```

Benjamin-Philip commented on erts/emulator/beam/erl_bif_persistent.c:
```diff
@@ -750,6 +664,174 @@ erts_init_persistent_dumping(void)
+    Eterm old_bucket;
+    long iterations_until_trap;
+    long max_iterations;
+#define PUT_TRAP_CODE                                            \
+    BIF_TRAP2(BIF_TRAP_EXPORT(BIF_persistent_term_put_2), c_p, state_mref, term)
+#define TRAPPING_COPY_TABLE_PUT(TABLE_DEST, OLD_TABLE, NEW_SIZE, COPY_TYPE, LOC_NAME) \
```
I've managed to pin point the previous trapping errors to here with some printf debugging. `TRAPPING_COPY_TABLE_PUT` stops just as when it runs `PUT_TRAP_CODE`.

I suspect the error is in calling `BIF_TRAP_EXPORT(BIF_persistent_term_put_2)` within `put_common`?

Benjamin-Philip reviewed (commented):
I've managed to unify trapping to a `put_common_trap`. However, I'm still facing some issues.

jhogberg commented:
> @jhogberg Do you plan not to pursue this right now in favour of your planned changes?

Sorry, I’ve had to put everything on ice and am not working much until October for personal reasons. 🙁

We can merge this PR once the issues are resolved, I can fix that come October.

jhogberg reviewed (commented):
The issues are probably related. Which of the arguments is it?

In the CI run it crashed during `persistent_term_SUITE:get_all_race/1`, suggesting that the problem is not specific to the new-key part of `put_common`, at least.

jhogberg commented on erts/emulator/beam/erl_bif_persistent.c:
```diff
@@ -306,140 +323,39 @@ static int persistent_term_put_2_ctx_bin_dtor(Binary *context_bin)
+    ErtsPersistentTermPutCommonResult result =
+        put_common(BIF_P, BIF_ARG_1, BIF_ARG_2, am_false);
+
+    #define HANDLE_PUT_COMMON_RESULT                            \
```
You can use those in whatever scope you wish, just change the arguments to match. `BIF_ERROR(BIF_P, BADARG);` can be rewritten to work in `put_common`s scope by writing it as `BIF_ERROR(c_p, BADARG);`, the same goes for `ERTS_BIF_YIELD3` and friends, just substitute `BIF_P` for `c_p`, `BIF_ARG_1` for `key`, `BIF_ARG_2` for `term`, and `BIF_ARG_3` for `new`, etc.

Benjamin-Philip commented:
@jhogberg, I pushed a working solution few days ago. Not sure why the 32-bit and cross-compile builds failed though. Everything else seems to be in order.

jhogberg commented on erts/emulator/beam/erl_bif_persistent.c:
```diff
@@ -543,130 +422,130 @@ BIF_RETTYPE persistent_term_erase_1(BIF_ALIST_1)
-        state_mref = erts_mk_magic_ref(&hp, &MSO(BIF_P), state_bin);
-        ctx = ERTS_MAGIC_BIN_DATA(state_bin);
+        TRAPPING_COPY_TABLE(TABLE_DEST, OLD_TABLE, NEW_SIZE, REHASH, LOC_NAME, ERASE_TRAP_CODE, BIF_P)
+if (is_internal_magic_ref(BIF_ARG_1) &&
```
Did something happen with the formatting here?

Benjamin-Philip commented:
About the formatting, I reformatted the functions I was working on with Emacs' `c-fill-paragraph`. I'll revert the changes. Do you have a configuration for better formatting?

jhogberg commented:
> About the formatting, I reformatted the functions I was working on with Emacs' `c-fill-paragraph`. I'll revert the changes. Do you have a configuration for better formatting?

I believe the problem is that, at least in `persistent_term_erase_1`, there's a missing `;` just before the formatting becomes weird.

> I never did find the cause, except that it was while trapping. The latest implementation does not crash.

Alright, we'll see if it crashes in the nightly runs. :-)

Benjamin-Philip commented:
@jhogberg Any further changes? Is `maint` still the right branch to merge into?

Benjamin-Philip commented:
> We generally keep ~moonshine~ moonshot ideas in our private issue tracker

Oh, I assure you it's not moonshine (or even moonshot) in the slightest. I think there is real value in better organizing persistent terms.

> One wrinkle is for each namespace to have its own options for read/update concurrency so the user can make trade-offs on whether they want the old hashtable version (faster reads) or the concurrent trie version (slightly slower reads, ridiculously fast updates).

Makes sense.

> but if you'd like to get started on an EEP that would be great.

I'll see if I can submit something in the coming months. No promises though. :)
